Rules cli manual (#192)
golemfactory/yagna-docs: provider-tutorials/provider-cli.md

`ya-provider` allows to fine-tune the config created with `golemsp settings` using commands `config`, `preset`, `profile`, and `exe-unit`.

Additionally, it enables configuration of the rules, certificate keystore and domain whitelist, which are used to control what kind of outbound traffic is allowed out of the VM containers run by the requestors on your machine.

### Keystore

> The provider has an embedded certificate keystore which is used to keep certificates which then can be used by specific Rules.

Reviewer: I believe this sentence carries no meaning after the change...

Author: What do you propose then?
This sentence says exactly what keystore now does: it only gathers all the certificates inside and nothing more. Then it can be used by specific rules to control i.e. outbound

Reviewer: maybe something along the lines of:

```suggestion
The provider has an embedded keystore, used to hold certificates. Specific rules can then reference those certificates to enable certain actions.
```

By default it contains only Golem public certificates for `AuditedPayload` and `Partner` rules.

> They allow to execute [example app](../requestor-tutorials/service-development/service-example-6-external-api-request.md) and apps from trusted by Golem creators (certificates allow to verify incoming _Demand_'s [Computational Payload Manifests](../requestor-tutorials/vm-runtime/computation-payload-manifest.md)).y default it contains only Golem public certificate which allows to execute [example app](../requestor-tutorials/service-development/service-example-6-external-api-request.md) and apps from trusted by Golem creators (certificates allow to verify incoming _Demand_'s [Computational Payload Manifests](../requestor-tutorials/vm-runtime/computation-payload-manifest.md)).

Reviewer: ```suggestion
They allow execution of the [example app](../requestor-tutorials/service-development/service-example-6-external-api-request.md) and apps from creators trusted by Golem (certificates allow to verify incoming _Demand_'s [Computational Payload Manifests](../requestor-tutorials/vm-runtime/computation-payload-manifest.md)).
```

### Rules

`ya-provider` offers a mechanism that allows to control the Rules for specific features i.e. Outbound Network.

#### Outbound Network feature

This feature defines rules and modes for Outbound Network.
It gives provider control over which entities are trusted and what level of Outbound Network access they have.

Outbound feature provides three rules:

1. `everyone`: defines the default outbound access for all requestors.
2. `audited-payload`: controls outbound access for manifests which are audited and signed by a trusted party.
   Trust is gained when provider sets this rule for root certificate of auditing organisation.
   Provider can have multiple `audited-payload` rules for different `X.509` certificates.
   To match this rule, requestor should get his manifest audited and signed by organisation which is trusted by provider.
   It doesn't matter from which requestor node request is made.
3. `partner`: allows outbound access for any manifests, as long as requestor node is a partner to trusted organisation.

>    Trust is gained when provider sets Partner rule for root certificate of trusted organisation.

Reviewer 2: ```suggestion
   Trust is gained when provider sets Partner rule for root `Golem` certificate of trusted organisation.
```

Reviewer 2: There is nowhere stated that Partner rule supports only `Golem`, and Audited Payload only `X.509`, so adding here "`Golem` certificate" will at least discourage setting Partner rule to `X.509` certificate.

3. `none`: no outbound traffic is allowed.

> **Warning**
> Remember, it is your responsibility as a provider to decide which entities you trust and the level of Outbound Network access you want to enable for them.

By default, the following rules and modes are set during installation:

- Everyone: Whitelist mode (every requestor can access outbound until domain is whitelisted).
- AuditedPayload: All mode (Golem root certificate is trusted).
- Partner: All mode (another Golem root certificate is trusted).

#### Listing Rules

To list currently set rules, you can use the `ya-provider rule list` command. By default, it displays the rules in a table format:

```
$ ya-provider rule list
Outbound status: enabled
┌──────────────────┬─────────────┬───────────────┬───────────────┐
│  rule            │  mode       │  certificate  │  description  │
├──────────────────┼─────────────┼───────────────┼───────────────┤
│  Everyone        │  whitelist  │               │               │
│  AuditedPayload  │  all        │  55e451bd     │               │
│  Partner         │  all        │  80c84b27     │               │
└──────────────────┴─────────────┴───────────────┴───────────────┘
```

> You can also retrieve the rules in JSON format by adding the `--json` flag:

Reviewer 2: Every `ya-provider` command printing table has `--json` flag, so maybe we can skip it?
A valid reason for keeping it is this issue with `certificate` id columns values, which can be different between `keystore list` and `rule list`, so it at least shows how to get full id.
Still it is unlikely anyone will face it.

Author: Yes, it can be shorter

Author: So I propose that it will stay here so that user can have easy way of finding big cert-id's so it can be used in keystore

The [Computational Payload Manifests](../requestor-tutorials/vm-runtime/computation-payload-manifest.md) embedded in the Demands can specify a list of URLs which may get called by the services running on a Provider. If the manifest declares requests to URLs in domains that are not whitelisted, it must come with a [signature and app author's public certificate](../requestor-tutorials/vm-runtime/computation-payload-manifest.md). By default, the domains `whitelist` consists of a curated set of public websites and APIs like github, dockerhub or public Ethereum nodes.

Run `ya-provider whitelist --help` to see possible subcommands

### Rules

> `ya-provider` offers a mechanism that allows to control the Rules for specific features i.e. Outbound Network.

Reviewer: ```suggestion
`ya-provider` offers a mechanism that enables control of the Rules governing specific features i.e. Outbound Network.
```

> #### Outbound Network feature

Reviewer: ```suggestion
#### Outbound Network feature
```

Why do we need to attach a title of "feature" to a feature? it's like saying "Master and Margaret book" instead of just "Master and Margaret" ...

This feature defines rules and modes for Outbound Network.
It gives provider control over which entities are trusted and what level of Outbound Network access they have.

Outbound feature provides three rules:

1. `everyone`: defines the default outbound access for all requestors.
2. `audited-payload`: controls outbound access for manifests which are audited and signed by a trusted party.

>    Trust is gained when provider sets this rule for root certificate of auditing organisation.

Reviewer: ```suggestion
   Trust is gained when a provider sets this rule for the root certificate of the auditing organisation.
```

but generally, this sentence sounds very awkward anyway...

Reviewer: perhaps this one would be better: 

```suggestion
   For the payload to be recognized as trusted, the provider must set this rule for the root certificate of the auditing organization
```

   Provider can have multiple `audited-payload` rules for different `X.509` certificates.

>    To match this rule, requestor should get his manifest audited and signed by organisation which is trusted by provider.

Reviewer: ```suggestion
   To match this rule, the requestor should get their manifest audited and signed by an organisation which is trusted by provider.
```

>    It doesn't matter from which requestor node request is made.

Reviewer: what request? do you mean a demand?

> 3. `partner`: allows outbound access for any manifests, as long as requestor node is a partner to trusted organisation.

Reviewer: ```suggestion
3. `partner`: allows outbound access for any manifests, as long as the requestor node is a partner to a trusted organisation.
```

>    Trust is gained when provider sets Partner rule for root certificate of trusted organisation.

Reviewer: very awkward-sounding... maybe consider something along the lines of:

```suggestion
   For the payload to be recognized as trusted, the provider must set a Partner rule for the root certificate of a trusted organization
```

>    Provider can have multiple `partner` rules for different [golem certificates](../requestor-tutorials/vm-runtime/golem-certificates.md).

Reviewer: ```suggestion
   A provider can have multiple `partner` rules, each for a different [golem certificate](../requestor-tutorials/vm-runtime/golem-certificates.md).
```

?

>    To match this rule, requestor should get his [node descriptor signed](../requestor-tutorials/vm-runtime/golem-certificates.md) by partner to trusted organisation.

Reviewer: ```suggestion
   To match this rule, a requestor should get their node descriptor signed](../requestor-tutorials/vm-runtime/golem-certificates.md) by a partner to a trusted organisation.
```

>    It doesn't matter which payload is being requested.

Reviewer: again, what is meant by "requested" here?

Every rule can operate in three different modes:

1. `all`: all domains listed in the manifest can be accessed for outbound traffic.
2. `whitelist`: only whitelisted domains can be accessed for outbound traffic.
3. `none`: no outbound traffic is allowed.

> **Warning**
> Remember, it is your responsibility as a provider to decide which entities you trust and the level of Outbound Network access you want to enable for them.

By default, the following rules and modes are set during installation:

> - Everyone: Whitelist mode (every requestor can access outbound until domain is whitelisted).

Reviewer: ```suggestion
- Everyone: Whitelist mode (every requestor can perform outbound connections as long as the target domain is whitelisted).
```

Outbound status: enabled
┌──────────────────┬─────────────┬───────────────┬───────────────┐
│  rule            │  mode       │  certificate  │  description  │
├──────────────────┼─────────────┼───────────────┼───────────────┤
│  Everyone        │  whitelist  │               │               │
│  AuditedPayload  │  all        │  55e451bd     │               │
│  Partner         │  all        │  80c84b27     │               │
└──────────────────┴─────────────┴───────────────┴───────────────┘
```

You can also retrieve the rules in JSON format by adding the `--json` flag:

```
$ ya-provider rule list --json
{
  "outbound": {
    "enabled": true,
    "everyone": "whitelist",
    "audited-payload": {
      "55e451bd1a2f43570a25052b863af1d527fe6fd4bfd1482fdb241596432477f20eb2b2f3801fb5c6cd785f1a03c43ccf71fd8cdf0a974d1296be2326b0824673": {
        "mode": "all",
        "description": ""
      }
    },
    "partner": {
      "80c84b2701126669966f46c1159cae89c58fb088e8bf94b318358fa4ca33ee56d8948511a397e5aba6aa5b88fff36f2541a91b133cde0fb816e8592b695c04c3": {
        "mode": "all",
        "description": ""
      }
    }
  }
}
```

#### Enabling and Disabling Outbound Traffic

> You can enable or disable the entire outbound traffic regardless of other rules settings by using the following commands:

Reviewer: can you indeed _enable_ "the entire network traffic" regardless of other rules? or just disable it?
